Neos.Flow/Classes/Core/Bootstrap.php

```diff
@@ -404,27 +403,54 @@ public function getObjectManager(): ObjectManagerInterface
      */
     protected function resolveRequestHandler(): RequestHandlerInterface
     {
-        if ($this->preselectedRequestHandlerClassName !== null && isset($this->requestHandlers[$this->preselectedRequestHandlerClassName])) {
-            $requestHandler = $this->requestHandlers[$this->preselectedRequestHandlerClassName];
+        if ($this->preselectedRequestHandlerClassName !== null) {
+            $requestHandler = $this->getRequestHandlerInstance($this->preselectedRequestHandlerClassName);
             if ($requestHandler->canHandleRequest()) {
                 return $requestHandler;
             }
         }
 
-        foreach ($this->requestHandlers as $requestHandler) {
+        /** @var RequestHandlerInterface|null $bestMatchingRequestHandler */
+        $bestMatchingRequestHandler = null;
+        $bestMatchingPriority = -9999;
+        foreach ($this->requestHandlerClassNames as $requestHandlerClassName) {
+            if ($bestMatchingPriority > $requestHandlerClassName::getPriority()) {
+                continue;
+            }
+
+            $requestHandler = $this->getRequestHandlerInstance($requestHandlerClassName);
             if ($requestHandler->canHandleRequest() > 0) {
-                $priority = $requestHandler->getPriority();
-                if (isset($suitableRequestHandlers[$priority])) {
+                $priority = $requestHandler::getPriority();
+                if ($bestMatchingPriority === $priority) {
                     throw new FlowException('More than one request handler with the same priority can handle the request, but only one handler may be active at a time!', 1176475350);
                 }
-                $suitableRequestHandlers[$priority] = $requestHandler;
+                $bestMatchingPriority = $priority;
+                $bestMatchingRequestHandler = $requestHandler;
             }
         }
-        if (empty($suitableRequestHandlers)) {
+
+        if (is_null($bestMatchingRequestHandler)) {
             throw new FlowException('No suitable request handler could be found for the current request. This is most likely a setup-problem, so please check your composer.json and/or try removing Configuration/PackageStates.php', 1464882543);
         }
-        ksort($suitableRequestHandlers);
-        return array_pop($suitableRequestHandlers);
+        return $bestMatchingRequestHandler;
+    }
+
+    /**
+     * @param class-string $className
+     * @return RequestHandlerInterface
+     */
+    private function getRequestHandlerInstance(string $className): RequestHandlerInterface
+    {
+        if (!in_array($className, $this->requestHandlerClassNames, true)) {
+            throw new FlowException(sprintf('A request handler "%s" was never registered', $className), 1719060769);
+        }
+
+        $requestHandler = new $className($this);
```

Comment:
im not so found of the implicit `public function __construct(Bootstrap $bootstrap)` part now ... at least its not declared in the `RequestHandlerInterface` ... maybe we should introduce a `public static function fromBootstrap(Bootstrap $bootstrap): self;` ?

Reply:
yeah good point, now that I mess with the interface anyways, makes sense. Added now